Source/Infrastructure/Integration/Kysect.Shreks.Integration.Google/GoogleTableAccessor.cs

```diff
@@ -1,23 +1,139 @@
+using Kysect.Shreks.Application.Abstractions.Google.Queries;
+using Kysect.Shreks.Core.Study;
 using Kysect.Shreks.Integration.Google.Sheets;
+using Kysect.Shreks.Integration.Google.Tools;
+using Kysect.Shreks.Integration.Google.Tools.Collection;
+using MediatR;
+using Microsoft.Extensions.Logging;
 
 namespace Kysect.Shreks.Integration.Google;
 
 public class GoogleTableAccessor : IGoogleTableAccessor
 {
+    private static readonly TimeSpan DelayBetweenSheetUpdates = TimeSpan.FromMinutes(1);
+
+    private static readonly ConcurrentHashSet<Guid> QueueUpdateSubjectCourseIds = new();
+    private static readonly ConcurrentHashSet<Guid> PointsUpdateSubjectCourseIds = new();
+
+    private static readonly SemaphoreSlim QueueUpdateSemaphore = new(1);
+    private static readonly SemaphoreSlim PointsUpdateSemaphore = new(1);
+
+    private static readonly SemaphoreSlim SpreadsheetCreationSemaphore = new(1);
+
     private readonly ISheet<CoursePoints> _pointsSheet;
     private readonly ISheet<StudentsQueue> _queueSheet;
+    private readonly ISheetManagementService _sheetManagementService;
+    private readonly IMediator _mediator;
+    private readonly ILogger<GoogleTableAccessor> _logger;
 
-    public GoogleTableAccessor(ISheet<CoursePoints> pointsSheet, ISheet<StudentsQueue> queueSheet)
+    public GoogleTableAccessor(
+        ISheet<CoursePoints> pointsSheet,
+        ISheet<StudentsQueue> queueSheet,
+        ISheetManagementService sheetManagementService,
+        IMediator mediator,
+        ILogger<GoogleTableAccessor> logger)
     {
         _pointsSheet = pointsSheet;
         _queueSheet = queueSheet;
+        _sheetManagementService = sheetManagementService;
+        _mediator = mediator;
+        _logger = logger;
+    }
+
+    public async Task UpdatePointsAsync(Guid subjectCourseId, CancellationToken token = default)
+    {
+        PointsUpdateSubjectCourseIds.Add(subjectCourseId);
+
+        await PointsUpdateSemaphore.WaitAsync(token);
+
+        IReadOnlyCollection<Guid> subjectCourseIds = PointsUpdateSubjectCourseIds.GetAndClearValues();
+        if (!subjectCourseIds.Any())
+        {
+            PointsUpdateSemaphore.Release();
+            return;
+        }
+        
+        foreach (var courseId in subjectCourseIds)
+        {
+            try
+            {
+                GetCoursePointsBySubjectCourse.Response response = await _mediator
+                    .Send(new GetCoursePointsBySubjectCourse.Query(courseId), token);
+
+                CoursePoints points = response.Points;
+
+                string spreadsheetId = await GetSpreadsheetIdAsync(courseId, token);
+                await _pointsSheet.UpdateAsync(spreadsheetId, points, token);
+
+                _logger.LogInformation($"Successfully updated points sheet of course {courseId}.");
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, $"Failed to update points sheet of course {courseId}.");
+            }
```

Comment:
При создании сообщений логера не стоит юзать интерполяцию строк.

```suggestion
            try
            {
                var query = new GetCoursePointsBySubjectCourse.Query(courseId);
                var response = await _mediator.Send(query, token);

                var points = response.Points;
                var spreadsheetId = await GetSpreadsheetIdAsync(courseId, token);
                await _pointsSheet.UpdateAsync(spreadsheetId, points, token);

                _logger.LogInformation("Successfully updated points sheet of course {CourseId}.", courseId);
            }
            catch (Exception e)
            {
                _logger.LogError(e, "Failed to update points sheet of course {CourseId}.", courseId);
            }
```

Reply:
А можно узнать почему так?

Reply:
На самом деле, строка которую ты передаёшь это темплейт. С правильной реализацией ты оставляешь информацию об имени параметра в темплейте, плюс не аллоцируешь много строк. 

https://www.youtube.com/watch?v=6zoMd_FwSwQ